Basic lowering generic function definitions.

alinas opened the pull request:
Resolve the specific type for the callee, to lower the proper specific function called.

alinas committed:
Refactor function definition lowering.

alinas committed:
Add basic lowering of generic function definitions.

alinas committed:
Basic lowering generic function definitions.

Resolve the specific type for the callee, to lower the proper specific
function called.

alinas commented:
Depends on #5015.

dwblaikie commented on toolchain/lower/testdata/function/generic/call_basic.carbon:
```diff
@@ -15,7 +15,8 @@ fn H[T:! type](x: T) {
 fn G[T:! type](x: T) -> T {
-  // TODO: call crashing.
+  H(x);
+  // TODO: Call crashes, see TODO in FunctionContext::GetValue()
   // H(T);
```
What would `H(T);` mean? (T is the type, right - but the H call needs a value, and the `H(x)` above shows that working) Is it just a "crashes instead of error-ing"?

alinas replied:
First off, yes, it shouldn't crash.
Second, I thought this is allowed. It's strange (and in this case useless), but the function could accept a type as the value and that would be legal. I inferred this from looking at existing test call.carbon that calls F(i32), but let me know if I misunderstood.

dwblaikie commented on toolchain/lower/file_context.cpp:
```diff
@@ -217,6 +217,16 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
+  // TODO nit: add is_symbolic() to type_id to forward to
+  // type_id.AsConstantId().is_symbolic(). Update call below too.
+  auto get_llvm_type = [&](SemIR::TypeId type_id) -> llvm::Type* {
+    return type_id.has_value() ? (type_id.AsConstantId().is_symbolic()
+                                      ? GetType(SemIR::GetTypeInSpecific(
+                                            sem_ir(), specific_id, type_id))
+                                      : GetType(type_id))
+                               : nullptr;
```
Maybe nested conditional operators are a bit confusing/worth writing out, maybe:
```
if (!type_id.has_value()) {
  return nullptr;
}
if (type_id.AsConstantId().is_symbolic()) {
  return GetType(SemIR::GetTypeInSpecific(...));
}
return GetType(type_id);
```
and/or I guess maybe popping the `GetType(` out could be valid:
```
return GetType(type_id.isConstantId().is_symbolic() ? SemIR::GetTypeInSpecific(...) : type_id);
```

dwblaikie commented on toolchain/lower/file_context.cpp:
```diff
@@ -247,6 +256,7 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
     return_param_id = function.return_slot_pattern_id;
     param_inst_ids.push_back(return_param_id);
   }
+
```
Drop unrelated formatting

dwblaikie commented on toolchain/lower/file_context.cpp:
```diff
@@ -255,8 +265,12 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
     }
-    auto param_type_id =
-        SemIR::GetTypeInSpecific(sem_ir(), specific_id, param_pattern.type_id);
+    auto param_type_id = param_pattern.type_id.AsConstantId().is_symbolic()
+                             ? SemIR::GetTypeInSpecific(sem_ir(), specific_id,
+                                                        param_pattern.type_id)
+                             : param_pattern.type_id;
```
This looks sort of like the smaller part of the above lambda, maybe worth factoring out/sharing.

dwblaikie commented on toolchain/lower/handle_call.cpp:
```diff
@@ -437,7 +437,14 @@ auto HandleInst(FunctionContext& context, SemIR::InstId inst_id,
   std::vector<llvm::Value*> args;
 
-  if (SemIR::ReturnTypeInfo::ForType(context.sem_ir(), inst.type_id)
+  auto inst_type_id =
+      callee_function.resolved_specific_id.has_value()
+          ? SemIR::GetTypeInSpecific(context.sem_ir(),
+                                     callee_function.resolved_specific_id,
+                                     inst.type_id)
+          : inst.type_id;
+
```
Another instance of this idiom.

alinas committed:
Address comments.

GetTypeInSpecific returns the given type id when given non-symbolic type. This simplifies the patch.
Also run auto-updater after earlier rebase.

danakj commented on toolchain/lower/file_context.cpp:
```diff
@@ -246,6 +266,8 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
     }
     auto param_type_id =
         SemIR::GetTypeInSpecific(sem_ir(), specific_id, param_pattern.type_id);
+    CARBON_CHECK(!param_type_id.AsConstantId().is_symbolic() &&
```
```suggestion
    CARBON_CHECK(!param_type_id.AsConstantId().is_symbolic(),
```

danakj commented on toolchain/lower/file_context.cpp:
```diff
@@ -308,20 +332,37 @@ auto FileContext::BuildFunctionDefinition(SemIR::FunctionId function_id)
+                                    llvm::Function* llvm_function,
+                                    SemIR::SpecificId specific_id) -> void {
+  const auto& body_block_ids = function.body_block_ids;
+  CARBON_DCHECK(llvm_function && !body_block_ids.empty());
```
```suggestion
  CARBON_DCHECK(llvm_function);
  CARBON_DCHECK(!body_block_ids.empty());
```
So you know which one failed

danakj replied:
Oh I see this came from https://github.com/carbon-language/carbon-lang/pull/5014, moving there.

danakj reviewed (commented):
Some little things I saw while reading

danakj commented on toolchain/lower/file_context.cpp:
```diff
@@ -85,14 +86,22 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
   }
+
+  // Lower function definitions for generics.
+  // TODO: this needs to be an iterator as new definitions can be added
```
Maybe this meant to say it can't be an iterator?
```suggestion
  // This can't be an iterator, and needs to use an index, as new definitions can be added
```

alinas committed:
Update DCHECK.

alinas committed:
Merge branch 'trunk' into func_def

alinas committed:
Small cleanups.

dwblaikie commented on toolchain/lower/function_context.h:
```diff
@@ -45,6 +46,8 @@ class FunctionContext {
 
   // Returns a value for the given instruction.
   auto GetValue(SemIR::InstId inst_id) -> llvm::Value* {
+    // TODO: if(specific_id_.has_value()) may need to update inst_id first.
```
could this be a CARBON_FATAL TODO sort of thing? (so if someone does write a test the fail, rather than getting some weird behavior later on?)

alinas replied:
The change here looks more involved to get this to work. I'd rather leave this as is in the PR and build on it in the next iteration, if you're ok with that.

alinas committed:
Merge branch 'trunk' into func_def
